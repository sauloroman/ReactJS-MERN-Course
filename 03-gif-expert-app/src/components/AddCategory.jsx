import { useState } from "react"

export const AddCategory = ({ onAddCategory }) => {

  const [category, setCategory] = useState('')

  const onSearchGifs = (e) => {
    e.preventDefault();
    if ( category.trim().length === 0 ) return;

    onAddCategory( category )
    setCategory('')
  }

  return (
    <form data-testid='form-add-category' onSubmit={ onSearchGifs } className="form">
      <div className="form__field">
        <input 
          value={category}
          onChange={ e => setCategory(e.target.value) }
          className="form__input"
          placeholder="Ingrese gifs a buscar"
          type="text" 
        />
        <i className='form__icon bx bx-search-alt-2'></i>
      </div>
    </form>
  )
}
