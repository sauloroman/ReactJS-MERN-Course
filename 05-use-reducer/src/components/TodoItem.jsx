export const TodoItem = ({ onDeleteTodo, onToggleTodo, id, title, description, isCompleted }) => {
  return (
    <div data-testid="todo" className={`todo ${ isCompleted && 'todo--completed'}`}>
      <div className="todo__info">
        <p className="todo__item-title">{title}</p>
        <p className="todo__description">{description}</p>
      </div>
      <div className="todo__controls">
        <button data-testid="button-toggle" onClick={ () => onToggleTodo( id ) } className="todo__button todo__button--complete">
          { isCompleted ? 'Habilitar' : 'Completar'}
        </button>
        <button data-testid="button-delete" onClick={() => onDeleteTodo( id )} className="todo__button todo__button--delete">Eliminar</button>
      </div>
    </div>
  )
}
