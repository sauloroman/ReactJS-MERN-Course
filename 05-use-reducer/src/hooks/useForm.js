import { useState } from "react"

export const useForm = ( initialState = {} ) => {

  const [formState, setFormState] = useState(initialState)

  const onInputChange = ({ target: { name, value }}) => {
    setFormState({
      ...formState,
      [name]: value
    })
  }

  const onResetForm = () => setFormState( initialState )

  return {
    formState,
    ...formState,

    onInputChange,
    onResetForm,
  }

}