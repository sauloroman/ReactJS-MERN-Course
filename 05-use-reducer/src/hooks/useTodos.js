import { useReducer } from "react"
import { todoReducer } from "../helpers/todo.reducer"

const init = () => JSON.parse(localStorage.getItem("todos")) || []

export const useTodos = () => {
    const [todos, dispatch] = useReducer( todoReducer, [], init )
  
    const onAddTodo = function( newTodo = {} ) {
      const actionAdd = {
        type: '[TODO] - Add New todo',
        payload: newTodo,
      }
      dispatch( actionAdd )
    }
  
    const onToggleTodo = function( todoId = '' ) {
      const actionToggle = {
        type: '[TODO] - Toggle todo',
        payload: todoId
      }
      dispatch( actionToggle )
    }
  
    const onDeleteTodo = function( todoId = '' ) {
      const actionDelete = {
        type: '[TODO] - Delete todo',
        payload: todoId,
      }
      dispatch( actionDelete )
    }

    return {
      todos,

      onAddTodo,
      onToggleTodo,
      onDeleteTodo,
    }
}