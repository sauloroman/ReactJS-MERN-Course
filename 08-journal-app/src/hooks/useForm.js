import { useEffect, useMemo, useState } from "react"

export const useForm = ( initialState = {}, formValidations = {} ) => {

  const [ formState, setFormState ] = useState( initialState )
  const [formValidation, setFormValidation] = useState({})

  useEffect(() => {
    createValidators()
  }, [formState])

  useEffect(() => {
    setFormState( initialState )
  }, [initialState])

  const isFormValid = useMemo(() => {
    for ( const formValue of Object.values(formValidation) ) {
      if ( formValue !== null ) return false
    }

    return true
  }, [formValidation])

  const onInputChange = ({ target }) => {

    const { name, value } = target

    setFormState({
      ...formState,
      [name]: value
    })

  }

  const onResetForm = () => setFormState( initialState )
  
  const createValidators = () => {

    const formCheckedValues = {}

    for( const formField of Object.keys( formValidations )) {
      const [ fn, errorMessage ] = formValidations[ formField ]
      formCheckedValues[`${formField}Valid`] = fn( formState[formField] ) ? null : errorMessage
    }
 
    setFormValidation({ ...formCheckedValues })

  }

  return {
    ...formState,
    ...formValidation,
    formState,
    isFormValid,

    onInputChange,
    onResetForm
  }

}