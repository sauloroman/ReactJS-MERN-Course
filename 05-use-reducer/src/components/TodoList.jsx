import { TodoItem } from './TodoItem'

export const TodoList = ({ onDeleteTodo, onToggleTodo, todos }) => {
  return (
    <div className="todos__list">
      <h2 className="todo__title">Todos Creados</h2>

      <div className="todos__list-grid">
        {
          todos?.map( todo => (
            <TodoItem 
              key={todo.id} 
              onDeleteTodo={onDeleteTodo} 
              onToggleTodo={onToggleTodo} 
              {...todo} 
            />
          ))
        }
      </div>
    </div>
  )
}
