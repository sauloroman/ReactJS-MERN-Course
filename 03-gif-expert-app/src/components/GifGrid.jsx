import { useFetchGifs } from "../hooks/useFetchGifs"
import { GifItem } from "./GifItem"

export const GifGrid = ({ onDeleteCategory, category }) => {

  const { isLoading, data } = useFetchGifs( category )

  return (
    <div className="gifs">
      <header className="gifs__header">
        <h2 className="category-name">{category[0] + category.slice(1).toLowerCase()}</h2>
        <button onClick={ () => onDeleteCategory(category) } className="btn">Eliminar</button>
      </header>
      {
        isLoading
        ? (<p>Cargando...</p>)
        : (<GifItem gifs={data} />)
      }
    </div>
  )
}
