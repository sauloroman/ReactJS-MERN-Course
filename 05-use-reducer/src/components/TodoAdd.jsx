import { v4 as uuidv4 } from 'uuid'
import { useForm } from '../hooks/useForm'

export const TodoAdd = ({ onAddTodo }) => {

  const { formState, title, description, onInputChange, onResetForm } = useForm({
    title: '',
    description: '',
  })

  const onCreateTodo = (e) => {
    e.preventDefault()

    if ( 
      title.trim().length === 0 || 
      description.trim().length === 0 
    ) return

    formState.id = uuidv4()
    formState.isCompleted = false

    onAddTodo(formState)
    onResetForm()
  }

  return (
    <div className="todo__add">
      <h2 className="todo__title">Agregar nuevo todo</h2>

      <form data-testid='form-todo' onSubmit={ onCreateTodo } action="#" className="todo__form">
        <div className="todo__form-field">
          <label htmlFor="todo-title" className="todo__form-label">Título</label>
          <input
            data-testid='todo-input-title'
            name='title'
            value={title}
            onChange={ onInputChange }
            id='todo-title'
            placeholder="Agrega un título"
            className="todo__form-input" 
            type="text" 
          />
        </div>
        <div className="todo__form-field">
          <label htmlFor="todo-description" className="todo__form-label">Descripción</label>
          <textarea
            data-testid='todo-input-description'
            className='todo__textarea'
            name='description'
            value={description}
            onChange={onInputChange}
            id="todo-description" 
            placeholder='Agrega una descripción '></textarea>
        </div>
        <div className="todo__form-box">
          <button className="todo__form-button">Agregar</button>
        </div>
      </form>

    </div>
  )
}
