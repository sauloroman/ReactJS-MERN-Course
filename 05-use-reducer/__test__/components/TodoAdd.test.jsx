import { fireEvent, render, screen } from "@testing-library/react"
import { TodoAdd } from "../../src/components/TodoAdd"
import { useForm } from "../../src/hooks/useForm"

jest.mock('../../src/hooks/useForm.js')

describe('Tests in TodoAdd Component', () => {

  const onAddTodoMock = jest.fn()

  const onInputChangeMock = jest.fn()
  const onResetFormMock = jest.fn()

  useForm.mockReturnValue({
    formState: { title: '', description: '' },
    title: '',
    description: '',
    onInputChange: onInputChangeMock,
    onResetForm: onResetFormMock
  })

  test('It must match with snapshot', () => {
    const { container } = render(<TodoAdd onAddTodo={onAddTodoMock} />)
    expect( container ).toMatchSnapshot()
  })

  test('It should not call onAddTodo function if the input fields are empty', () => {

    render(<TodoAdd onAddTodo={onAddTodoMock} />)

    const formElement = screen.getByTestId('form-todo')
    fireEvent.submit( formElement )

    expect( onAddTodoMock ).not.toHaveBeenCalled()
    expect( onAddTodoMock ).toHaveBeenCalledTimes(0)

  })

  test('it should call onAddTodo function if the information is set', () => {

    const todoTest = {
      title: 'Title test 1',
      description: 'Description test 1'
    }

    useForm.mockReturnValue({
      formState: todoTest,
      title: todoTest.title,
      description: todoTest.description,
      onInputChange: onInputChangeMock,
      onResetForm: onResetFormMock
    })

    render(<TodoAdd onAddTodo={onAddTodoMock} />)

    const form = screen.getByTestId('form-todo')
    fireEvent.submit( form )

    expect( onAddTodoMock ).toHaveBeenCalledWith({
      ...todoTest,
      id: expect.any( String ),
      isCompleted: false,
    })

  })

})