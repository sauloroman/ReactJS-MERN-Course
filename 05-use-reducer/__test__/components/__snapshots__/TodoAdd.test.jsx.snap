// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in TodoAdd Component It must match with snapshot 1`] = `
<div>
  <div
    class="todo__add"
  >
    <h2
      class="todo__title"
    >
      Agregar nuevo todo
    </h2>
    <form
      action="#"
      class="todo__form"
      data-testid="form-todo"
    >
      <div
        class="todo__form-field"
      >
        <label
          class="todo__form-label"
          for="todo-title"
        >
          Título
        </label>
        <input
          class="todo__form-input"
          data-testid="todo-input-title"
          id="todo-title"
          name="title"
          placeholder="Agrega un título"
          type="text"
          value=""
        />
      </div>
      <div
        class="todo__form-field"
      >
        <label
          class="todo__form-label"
          for="todo-description"
        >
          Descripción
        </label>
        <textarea
          class="todo__textarea"
          data-testid="todo-input-description"
          id="todo-description"
          name="description"
          placeholder="Agrega una descripción "
        />
      </div>
      <div
        class="todo__form-box"
      >
        <button
          class="todo__form-button"
        >
          Agregar
        </button>
      </div>
    </form>
  </div>
</div>
`;
